amethyst_input/src/sdl_events_system.rs

```diff
@@ -0,0 +1,253 @@
+use std::fmt;
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::path::PathBuf;
+
+use amethyst_core::shrev::EventChannel;
+use amethyst_core::specs::prelude::{Resources, RunNow, SystemData, Write};
+use sdl2;
+use sdl2::controller::{AddMappingError, Axis, Button, GameController};
+use sdl2::event::Event;
+use sdl2::{EventPump, GameControllerSubsystem, Sdl};
+
+use super::controller::{ControllerAxis, ControllerButton, ControllerEvent};
+use super::{InputEvent, InputHandler};
+
+#[derive(Debug)]
+pub enum SdlSystemError {
+    ContextInit(String),
+    ControllerSubsystemInit(String),
+    AddMappingError(AddMappingError),
+}
+
+impl fmt::Display for SdlSystemError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            SdlSystemError::ContextInit(ref msg) => write!(f, "Failed to initialize SDL: {}", msg),
+            SdlSystemError::ControllerSubsystemInit(ref msg) => {
+                write!(f, "Failed to initialize SDL controller subsystem: {}", msg)
+            }
+            SdlSystemError::AddMappingError(ref err) => {
+                write!(f, "Failed to load controller mappings: {}", err)
+            }
+        }
+    }
+}
+
+pub enum ControllerMappings {
+    FromPath(PathBuf),
+    FromString(String),
+}
+
+pub struct SdlEventsSystem<AX, AC>
+where
+    AX: Hash + Eq,
+    AC: Hash + Eq,
+{
+    #[allow(dead_code)]
+    sdl_context: Sdl,
+    event_pump: Option<EventPump>,
+    controller_subsystem: GameControllerSubsystem,
+    /// Vector of opened controllers and their corresponding joystick indices
+    opened_controllers: Vec<(u32, GameController)>,
+    marker: PhantomData<(AX, AC)>,
+}
+
+type SdlEventsData<'a, AX, AC> = (
+    Write<'a, InputHandler<AX, AC>>,
+    Write<'a, EventChannel<InputEvent<AC>>>,
+);
+
+impl<'a, AX, AC> RunNow<'a> for SdlEventsSystem<AX, AC>
+where
+    AX: Hash + Eq + Clone + Send + Sync + 'static,
+    AC: Hash + Eq + Clone + Send + Sync + 'static,
+{
+    fn run_now(&mut self, res: &'a Resources) {
+        let (mut handler, mut output) = SdlEventsData::fetch(res);
+
+        let mut event_pump = self.event_pump.take().unwrap();
```

Comment:
Not sure I understand why we do this `take()` and put it back dance.  Can't we just call [`as_mut()`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.as_mut)?

Reply:
This is needed because I need `&mut self` for `handle_sdl_event`, but if I would just borrow the event pump, I would have two mutable refs. Luckily, it's unlikely you will ever need that `event_pump` inside the event handlers, so I just take it out to leave `self` not referenced by the loop iterator.